File: src/Tests/Behavioral/NamedFuncTypeMapParam/NamedFuncTypeMapParam.go
// Guards a methodless named func TYPE whose parameter list carries a named MAP parameter, used as
// a function parameter so the type collapses to its base C# delegate. The delegate's parameter
// types are rendered by convertToCSFullTypeName's func-handler: extractTypes already returns each
// param in C# form (`map<@string, ж<Node>>`), so a SECOND convertToCSTypeName pass re-fed that
// through the `map<` arm's splitMapKeyValue, mis-parsing it to `map<@string, ж<Node>, >` — a spurious
// trailing empty type arg (CS1031 "Type expected"). Same shape as go/ast's `type Importer
// func(imports map[string]*Object, path string) (pkg *Object, err error)` used by NewPackage.
package main

import "fmt"

type Node struct {
	name string
}

// methodless named func type: named MAP param + a plain param + named results.
type Importer func(imports map[string]*Node, path string) (pkg *Node, err error)

// importer's type collapses to the base delegate here (the double-conversion site).
func newPackage(importer Importer, files map[string]*Node) *Node {
	n, _ := importer(files, "root")
	return n
}

// a named function of the same signature — passed as the Importer value.
func lookup(imports map[string]*Node, path string) (pkg *Node, err error) {
	return imports["a"], nil
}

func main() {
	files := map[string]*Node{"a": {name: "alpha"}}
	pkg := newPackage(lookup, files)
	fmt.Println(pkg.name) // alpha
}

File: src/Tests/Behavioral/NamedFuncTypeMapParam/NamedFuncTypeMapParam.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

[GoType] partial struct Node {
    internal @string name;
}

// type Importer is a methodless func type — rendered inline as its base delegate

internal static ж<Node> newPackage(Func<map<@string, ж<Node>>, @string, (ж<Node>, error)> importer, map<@string, ж<Node>> files) {
    var (n, _) = importer(files, "root"u8);
    return n;
}

internal static (ж<Node> pkg, error err) lookup(map<@string, ж<Node>> imports, @string path) {
    ж<Node> pkg = default!;
    error err = default!;

    return (imports["a"u8], default!);
}

internal static void Main() {
    var files = new map<@string, ж<Node>>{["a"u8] = Ꮡ(new Node(name: "alpha"u8))};
    var pkg = newPackage(new Func<map<@string, ж<Node>>, @string, (ж<Node>, error)>(lookup), files);
    fmt.Println((~pkg).name);
}

} // end main_package
